src/main/java/org/reaktivity/nukleus/kafka/internal/stream/ClientStreamFactory.java

```diff
@@ -345,43 +346,49 @@ private void doKafkaData(
         final OctetsFW payload,
-        final Long2LongHashMap fetchOffsets)
+        final Long2LongHashMap fetchOffsets,
+        final OctetsFW messageKey)
     {
         final DataFW data = dataRW.wrap(writeBuffer, 0, writeBuffer.capacity())
                 .streamId(targetId)
                 .groupId(0)
                 .padding(padding)
-                .payload(p -> p.set(payload.buffer(), payload.offset(), payload.sizeof()))
-                .extension(e -> e.set(visitKafkaDataEx(fetchOffsets)))
+                .payload(payload)
+                .extension(e -> e.set(visitKafkaDataEx(fetchOffsets, messageKey)))
                 .build();
 
         target.accept(data.typeId(), data.buffer(), data.offset(), data.sizeof());
     }
 
     private Flyweight.Builder.Visitor visitKafkaDataEx(
-        Long2LongHashMap fetchOffsets)
+        Long2LongHashMap fetchOffsets,
+        final OctetsFW messageKey)
     {
         return (b, o, l) ->
         {
-            return dataExRW.wrap(b, o, l)
-                           .fetchOffsets(a ->
-                           {
-                               if (fetchOffsets.size() == 1)
-                               {
-                                   final long offset = fetchOffsets.values().iterator().nextValue();
-                                   a.item(p -> p.set(offset));
-                               }
-                               else
-                               {
-                                   int partition = -1;
-                                   while (fetchOffsets.get(++partition) != fetchOffsets.missingValue())
-                                   {
-                                       final long offset = fetchOffsets.get(partition);
-                                       a.item(p -> p.set(offset));
-                                   }
-                               }
-                           })
-                           .build()
+            dataExRW.wrap(b, o, l)
+                    .fetchOffsets(a ->
+                    {
+                        if (fetchOffsets.size() == 1)
+                        {
+                            final long offset = fetchOffsets.values().iterator().nextValue();
+                            a.item(p -> p.set(offset));
+                        }
+                        else
+                        {
+                            int partition = -1;
+                            while (fetchOffsets.get(++partition) != fetchOffsets.missingValue())
+                            {
+                                final long offset = fetchOffsets.get(partition);
+                                a.item(p -> p.set(offset));
+                            }
+                        }
+                    });
+            if (messageKey != null)
+            {
+                dataExRW.messageKey(messageKey.buffer(), messageKey.offset(), messageKey.sizeof());
+            }
+            return dataExRW.build()
```

Comment:
if `dataExRW.messageKey(null)` works using `KafkaDataEx.Builder.messageKey(OctetsFW messageKey)` method, then let's use `dataExRW.messageKey(messageKey)` unconditionally here and keep the builder flow intact.

Reply:
thanks, done